fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/PhysicalDistributionAggOptRule.java

```diff
@@ -0,0 +1,130 @@
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class PhysicalDistributionAggOptRule implements TreeRewriteRule {
+    @Override
+    public OptExpression rewrite(OptExpression root, TaskContext taskContext) {
+        if (ConnectContext.get().getSessionVariable().isEnableQueryCache()) {
+            return root;
+        }
+        if (ConnectContext.get().getSessionVariable().isEnableSortAggregate()) {
+            root.getOp().accept(new UseSortAGGRule(), root, null);
+            return root;
+        }
+        if (ConnectContext.get().getSessionVariable().isEnablePerBucketComputeOptimize()) {
+            root.getOp().accept(new UsePerBucketOptimizeRule(), root, null);
+            return root;
+        }
+        return root;
+    }
+
+    private static class NoopVisitor extends OptExpressionVisitor<Void, Void> {
+        @Override
+        public Void visit(OptExpression optExpression, Void context) {
+            for (OptExpression opt : optExpression.getInputs()) {
+                opt.getOp().accept(this, opt, context);
+            }
+            return null;
+        }
+    }
+
+    private static class UsePerBucketOptimizeRule extends NoopVisitor {
+        @Override
+        public Void visitPhysicalHashAggregate(OptExpression optExpression, Void context) {
+            if (optExpression.getInputs().get(0).getOp().getOpType() != OperatorType.PHYSICAL_OLAP_SCAN) {
+                return visit(optExpression, context);
+            }
+
+            PhysicalOlapScanOperator scan = (PhysicalOlapScanOperator) optExpression.getInputs().get(0).getOp();
+            PhysicalHashAggregateOperator agg = (PhysicalHashAggregateOperator) optExpression.getOp();
+
+            // Now we only support one-stage AGG for per-bucket optimize
+            if (!agg.getType().isGlobal() || agg.getGroupBys().isEmpty()) {
+                return null;
+            }
+
+            agg.setUsePerBucketOptmize(true);
+            scan.setNeedOutputChunkByBucket(true);
+            return null;
+        }
+    }
+
+    private static class UseSortAGGRule extends NoopVisitor {
+        @Override
+        public Void visitPhysicalHashAggregate(OptExpression optExpression, Void context) {
+            if (optExpression.getInputs().get(0).getOp().getOpType() != OperatorType.PHYSICAL_OLAP_SCAN) {
+                return visit(optExpression, context);
+            }
+
+            PhysicalOlapScanOperator scan = (PhysicalOlapScanOperator) optExpression.getInputs().get(0).getOp();
+            PhysicalHashAggregateOperator agg = (PhysicalHashAggregateOperator) optExpression.getOp();
+
+            // Now we only support one-stage AGG
+            // TODO: support multi-stage AGG
+            if (!agg.getType().isGlobal() || agg.getGroupBys().isEmpty()) {
+                return null;
+            }
+
+            // the same key in multi partition are not in the same tablet
+            if (scan.getSelectedPartitionId().size() > 1) {
```

Comment:
It  seems that we can use per-bucket algorithms(either hash-agg or sort-agg)to process 1st-stage of  multi-stage agg.

Reply:
per-bucket optimize may have bad case for (1st-stage). it may cause more rpc call. (output hash-table-multi times).